processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -0,0 +1,1321 @@
+        value = value == null ? "" : value;
+        writer.write(value);
+
+        for (int i = 0; i < adapters.size(); i++) {
+          IndexMerger.DimValueConverter converter = converters[i];
+          if (converter != null) {
+            converter.convert(value, cardinality);
+          }
+        }
+        ++cardinality;
+      }
+
+      progress.progressSection(section, "dim[" + dimension + "] cardinality: " + cardinality);
+      dimensionCardinalities.put(dimension, cardinality);
+      writer.close();
+
+      if (cardinality == 0) {
+        progress.progressSection(section, String.format("Skipping [%s], it is empty!", dimension));
+        dimensionSkipFlag.add(true);
+        continue;
+      }
+      dimensionSkipFlag.add(false);
+
+      // make the conversion
+      for (int i = 0; i < adapters.size(); ++i) {
+        IndexMerger.DimValueConverter converter = converters[i];
+        if (converter != null) {
+          dimConversions.get(i).put(dimension, converters[i].getConversionBuffer());
+        }
+      }
+    }
+    progress.stopSection(section);
+  }
+
+  private void mergeCapabilities(
+      final List<IndexableAdapter> adapters,
+      final List<String> mergedDimensions,
+      final Map<String, ValueType> metricsValueTypes,
+      final Map<String, String> metricTypeNames,
+      final List<ColumnCapabilitiesImpl> dimCapabilities
+  )
+  {
+    final Map<String, ColumnCapabilitiesImpl> capabilitiesMap = Maps.newHashMap();
+    for (IndexableAdapter adapter : adapters) {
+      for (String dimension : adapter.getDimensionNames()) {
+        ColumnCapabilitiesImpl mergedCapabilities = capabilitiesMap.get(dimension);
+        if (mergedCapabilities == null) {
+          mergedCapabilities = new ColumnCapabilitiesImpl();
+          mergedCapabilities.setType(ValueType.STRING);
+        }
+        capabilitiesMap.put(dimension, mergedCapabilities.merge(adapter.getCapabilities(dimension)));
+      }
+      for (String metric : adapter.getMetricNames()) {
+        ColumnCapabilitiesImpl mergedCapabilities = capabilitiesMap.get(metric);
+        ColumnCapabilities capabilities = adapter.getCapabilities(metric);
+        if (mergedCapabilities == null) {
+          mergedCapabilities = new ColumnCapabilitiesImpl();
+        }
+        capabilitiesMap.put(metric, mergedCapabilities.merge(capabilities));
+        metricsValueTypes.put(metric, capabilities.getType());
+        metricTypeNames.put(metric, adapter.getMetricType(metric));
+      }
+    }
+    for (String dim : mergedDimensions) {
+      dimCapabilities.add(capabilitiesMap.get(dim));
+    }
+  }
+
+  private <T extends Comparable> ArrayList<T> mergeIndexed(final List<Iterable<T>> indexedLists)
+  {
+    Set<T> retVal = Sets.newTreeSet(Ordering.<T>natural().nullsFirst());
+
+    for (Iterable<T> indexedList : indexedLists) {
+      for (T val : indexedList) {
+        retVal.add(val);
+      }
+    }
+    return Lists.newArrayList(retVal);
+  }
+}
```

Comment:
This file reads more like IndexMaker in how it breaks up the indexing sections, which I like :)

I find this code much easier to read